### PR TITLE
Trackers: context instance discrimination

### PR DESCRIPTION
**Pull request by sdirosa**
Our contexts are uniquely identified by combining their `_type` and `id` attributes. 

For validation purposes this is not enough. We need to track contexts at instance level to know whether a certain item in a Location Stack is the same instance as an identical one (looking at the _type+id combination) in another Location Stack. Think of generic components tracking the same Context, but being rendered multiple times in different places of the UI.

Until now this was achieved in React Core. There was some custom code to extend the Core Factories with an extra attribute to achieve what we described above.

Since this mechanism is actually going to be generic, and eventually also used in Browser Tracker, I've moved it to the generic logic to the TS schema and modified the generator script.

**TS schema generator script changes**
The generator script now produces an `AbstractContext` definition that has a new discriminatory property called `__instance_id`. This property is not sent to the Collector but exists solely client-side for validation purposes, hence the double underscore as prefix.

I've also modified how the generator produces the `ContextFactories` of Core Tracker so it populates the `instance_id` property with the help of the Core Tracker `generateUUID` API.

**TL:DR** This change ensures that all Contexts created via the Core Factories contain now an internal attribute we can use to compare them, even if they have the same _type and id.

### CODE REVIEW

**Comment by sdirosa** [comment]
@borft Could you double check the generator script changes?

https://github.com/objectiv/objectiv-analytics/pull/641/files#diff-4fb779d5f8787394aa890917a2c573237a347f1053dd573ebe975ce8c2427119

File: tracker/core/react/src/common/factories/makeContentContext.ts [hunk 1]
```diff
@@ -1,15 +0,0 @@
...
-import { makeContentContext as coreMakeContentContext } from '@objectiv/tracker-core';
-import { LocationContext } from '../../types';
-import { makeLocationContext } from './makeLocationContext';
-
-/**
- * A utility to factor a new uniquely identifiable OverlayContext.
- */
-export const makeContentContext = (
-  props: Parameters<typeof coreMakeContentContext>[0]
-): LocationContext<ContentContext> => makeLocationContext<ContentContext>(coreMakeContentContext(props));
```
Comment (sdirosa) [inline tracker/core/react/src/common/factories/makeContentContext.ts:15]: All of these factories existed for the sole purpose of adding the discriminatory attribute, since that now is generically part of the Core Factories we can throw all of these away.

**Review by borft — APPROVED** [review]
lgtm, one small question

File: tracker/core/utilities/src/generator.js [hunk 2]
```diff
@@ -523,6 +523,17 @@ Object.entries(contexts).forEach(([context_type, context]) => {
       const discriminator = camelToUnderscore(context_type.replace('Abstract', ''));
       properties[DISCRIMINATING_PROPERTY_PREFIX + discriminator] = { discriminator: true };
     }
+
+    // Add __instance_id discriminator to AbstractContext
+    if (context_type === 'AbstractContext') {
+      properties[DISCRIMINATING_PROPERTY_PREFIX.repeat(2) + 'instance_id'] = {
+        discriminator: 'generateUUID()',
+        description: 'A unique identifier to discriminate Context instances across Location Stacks.',
```
[the remark this answers is not in the document]
Comment (sdirosa) [reply tracker/core/utilities/src/generator.js:531]: Yes. 

`description` and `type` end up in the Abstract class, e.g.
```
export abstract class AbstractContext {
  /**
   * A unique identifier to discriminate Context instances across Location Stacks.
   */
  __instance_id: string;

``` 

Value is used in the factory implementation:
```
export const makeApplicationContext = (props: { id: string }): ApplicationContext => ({
  __instance_id: generateUUID(),
  __global_context: true,
  _type: 'ApplicationContext',
  id: props.id,
});
```
